src/garage/torch/distributions/tanh_normal.py

```diff
@@ -0,0 +1,290 @@
+"""A Gaussian distribution with tanh transformation."""
+from collections import namedtuple
+
+import torch
+from torch.distributions import Normal
+from torch.distributions.independent import Independent
+
+
+class TanhNormal(torch.distributions.Distribution):
+    """A gaussian distribution that has a tanh transformation applied to it.
+
+    Algorithms like SAC and Pearl use this transformed distribution.
+    It can be thought of as a distribution of X where
+        Y ~ N(mean, cov)
+        X ~ tanh(Y)
+
+    Args:
+        mean (torch.tensor): The mean of this distribution.
+        std (torch.tensor): The stdev of this distribution.
+
+    """
+
+    def __init__(self, mean, std):
+        self._normal = Independent(Normal(mean, std), 1)
+        super().__init__()
+
+    def log_prob(self, value, pre_tanh_value=None, epsilon=1e-6):
+        """The log likelihood of a sample on the this Tanh Distribution.
+
+        Args:
+            value (torch.Tensor): The sample whose loglikelihood is being
+                computed.
+            pre_tanh_value (torch.Tensor): The value prior to having the tanh
+                function applied to it but after it has been sampled from the
+                normal distribution.
+            epsilon (float): Stabilization coefficient.
+
+        Notes:
+            - when pre_tanh_value is None, an estimate is made of what the
+              value is. This leads to a worse estimation of the log_prob.
+              If the value being used is collected from functions like
+              `sample` and `rsample`, one can instead use functions like
+              `sample_return_pre_tanh_value` or
+              `rsample_return_pre_tanh_value`
+
+
+        Returns:
+            torch.Tensor: The log likelihood of value on the distribution.
+
+        """
+        # pylint: disable=arguments-differ
+        if pre_tanh_value is None:
+            pre_tanh_value = torch.log((1 + value) / (1 - value)) / 2
+        norm_lp = self._normal.log_prob(pre_tanh_value)
+        ret = (norm_lp - torch.sum(
+            torch.log(self._clip_but_pass_gradient((1. - value**2)) + epsilon),
+            axis=-1))
+        return ret
+
+    def sample_return_pre_tanh_value(self, sample_shape=torch.Size()):
```

Comment:
perhaps `sample_unwrapped` or `sample_with_pretanh_value`?

i looked in your SAC PR and couldn't find any usage of this function. What is it for?

Would it be sufficient to provide `sample_unwrapped` (which samples the wrapped distribution) and `sample` (which provides the wrapped samples)? does anyone use both values at once in the differentiable path?

Reply:
I'll change the names but keep the function. There is a chance that the function is needed later on (not in SAC however)

Reply:
if the function has no known use or user then it should be deleted.

Reply:
I've removed it.

Reply:
What was the original use for it? When did you decide to remove it? (previously i thought it was essential to SAC working).

Reply:
no this function isn't necessary to making sac work. `rsample_with_pretanh_value` is necessary to SAC working.

Reply:
oh, in that case i take back what i said. yes you want the API to be symmetric so you should have both methods